sql/core/src/main/scala/org/apache/spark/sql/execution/MetadataOnlyOptimizer.scala

```diff
@@ -0,0 +1,171 @@
+      // Moreover, multiple-distinct has been rewritted into it by RewriteDistinctAggregates.
+      true
+    } else {
+      aggregateExpressions.forall { agg =>
+        if (agg.isDistinct) {
+          true
+        } else {
+          // If function can be evaluated on just the distinct values of a column, it can be used
+          // by metadata-only optimizer.
+          agg.aggregateFunction match {
+            case max: Max => true
+            case min: Min => true
+            case hyperLog: HyperLogLogPlusPlus => true
+            case _ => false
+          }
+        }
+      }
+    }
+  }
+
+  private def convertLogicalToMetadataOnly(
+      project: LogicalPlan,
+      filter: Option[Expression],
+      logical: LogicalRelation,
+      files: HadoopFsRelation): LogicalPlan = {
+    val attributeMap = logical.output.map(attr => (attr.name, attr)).toMap
+    val partitionColumns = files.partitionSchema.map { field =>
+      attributeMap.getOrElse(field.name, throw new AnalysisException(
+        s"Unable to resolve ${field.name} given [${logical.output.map(_.name).mkString(", ")}]"))
+    }
+    val projectSet = filter.map(project.references ++ _.references).getOrElse(project.references)
+    if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+      val selectedPartitions = files.location.listFiles(filter.map(Seq(_)).getOrElse(Seq.empty))
+      val valuesRdd = sparkSession.sparkContext.parallelize(selectedPartitions.map(_.values), 1)
+      val valuesPlan = LogicalRDD(partitionColumns, valuesRdd)(sparkSession)
+      valuesPlan
+    } else {
+      logical
+    }
+  }
+
+  private def convertCatalogToMetadataOnly(
+      project: LogicalPlan,
+      filter: Option[Expression],
+      relation: CatalogRelation): LogicalPlan = {
+    val attributeMap = relation.output.map(attr => (attr.name, attr)).toMap
+    val partitionColumns = relation.catalogTable.partitionColumnNames.map { column =>
+      attributeMap.getOrElse(column, throw new AnalysisException(
+        s"Unable to resolve ${column} given [${relation.output.map(_.name).mkString(", ")}]"))
+    }
+    val projectSet = filter.map(project.references ++ _.references).getOrElse(project.references)
+    if (projectSet.subsetOf(AttributeSet(partitionColumns))) {
+      val partitionColumnDataTypes = partitionColumns.map(_.dataType)
+      val partitionValues = catalog.listPartitions(relation.catalogTable.identifier)
+        .map { p =>
+          InternalRow.fromSeq(
+            partitionColumns.map(a => p.spec(a.name)).zip(partitionColumnDataTypes).map {
+              case (rawValue, dataType) => Cast(Literal(rawValue), dataType).eval(null)
+            })
+        }
+      val valuesRdd = sparkSession.sparkContext.parallelize(partitionValues, 1)
+      val valuesPlan = LogicalRDD(partitionColumns, valuesRdd)(sparkSession)
+      valuesPlan
+    } else {
+      relation
+    }
+  }
+
+  private def convertToMetadataOnly(plan: LogicalPlan): LogicalPlan = plan match {
```

Comment:
We need to think about it more carefully, i.e. how can the partition information propagate up from table relation?
It's obvious that `Filter` can retain all partition information, but for others, it's not trivial to explain.

Since this PR definitely need more people to review, how about we only handle `Filter` for now and improve it later? Then it's easier for other people to review and get this PR in. Thanks!